File: src/schema/application/grid_filters.ts
import { boolean, index, jsonb, pgTable, text } from 'drizzle-orm/pg-core';
import * as path from 'path';
import {
  fileRegex,
  system_fields,
} from '@dna-platform/crdt-lww-postgres/build/schema/system';
import { table as contacts } from './contacts';
const filename = path.basename(__filename).replace(fileRegex, '');

const fields = {
  name: text(),
  grid_id: text(),
  link: text().default(''),
  is_current: boolean().default(false),
  is_default: boolean().default(false),
  contact_id: text().references(() => (contacts as Record<string, any>).id),
  account_organization_id : text(),
  entity: text(),
  columns: jsonb("columns").default([]),
  groups: jsonb("groups").default([]),
  sorts: jsonb("sorts").default([]),
  // timestamp: timestamp('timestamp', { withTimezone: true }),
  default_sorts:jsonb("default_sorts").default([]),
  advance_filters: jsonb("advance_filters").default([]),
  group_advance_filters: jsonb("group_advance_filters").default([]),
  filter_groups: jsonb("filter_groups").default([]),
};

export const table = pgTable(
  filename,
  {
    ...system_fields,
    ...fields,
  },
  (table: Record<string, any>) => ({
    ...Object.keys({ ...system_fields, ...fields }).reduce((acc, field) => {
      const searchable_fields = [
        ...Object.keys(system_fields),
        'name',
        'grid_id',
        'link',
        'is_current',
        'is_default',
        'contact_id',
        'entity',
      ];
      if (!searchable_fields.includes(field)) return acc;
      const index_name = `grid_filters_${field}_idx`;
      if (index_name.includes('_time')) return acc;
      return {
        ...acc,
        [index_name]: index(index_name).on(table[field]),
      };
    }, {}),
  }),
);

File: bin/store/src/database/schema/tables/grid_filters.rs
use crate::define_table_schema;
use crate::generated::schema::generator::diesel_schema_definition::{
    types::*, DieselTableDefinition,
};
use crate::{system_fields, system_foreign_keys, system_indexes};

/**
 * Auto-converted from TypeScript schema source.
 */
pub struct GridFiltersTable;

define_table_schema! {
    hypertable: false,
    fields: {
        // System fields - common across all tables ( REQUIRED )
        system_fields!(),

        name: nullable(text()),
        grid_id: nullable(text()),
        link: nullable(text()), default "",
        is_current: nullable(boolean()), default "false",
        is_default: nullable(boolean()), default "false",
        contact_id: nullable(text()),
        account_organization_id: nullable(text()),
        entity: nullable(text()),
        columns: nullable(jsonb()), default "'[]'::jsonb",
        groups: nullable(jsonb()), default "'[]'::jsonb",
        sorts: nullable(jsonb()), default "'[]'::jsonb",
        default_sorts: nullable(jsonb()), default "'[]'::jsonb",
        advance_filters: nullable(jsonb()), default "'[]'::jsonb",
        group_advance_filters: nullable(jsonb()), default "'[]'::jsonb",
        filter_groups: nullable(jsonb()), default "'[]'::jsonb",
    },
    indexes: {
        // System field indexes ( REQUIRED )
        system_indexes!("grid_filters"),
        // Custom table-specific indexes
        idx_grid_filters_name: {
            columns: ["name"],
            unique: false,
            type: "btree"
        },
        idx_grid_filters_grid_id: {
            columns: ["grid_id"],
            unique: false,
            type: "btree"
        },
        idx_grid_filters_link: {
            columns: ["link"],
            unique: false,
            type: "btree"
        },
        idx_grid_filters_is_current: {
            columns: ["is_current"],
            unique: false,
            type: "btree"
        },
        idx_grid_filters_is_default: {
            columns: ["is_default"],
            unique: false,
            type: "btree"
        },
        idx_grid_filters_contact_id: {
            columns: ["contact_id"],
            unique: false,
            type: "btree"
        },
        idx_grid_filters_entity: {
            columns: ["entity"],
            unique: false,
            type: "btree"
        },
    },
    foreign_keys: {
        // System field foreign keys ( REQUIRED )
        system_foreign_keys!("grid_filters"),
        fk_grid_filters_contact_id: { columns: ["contact_id"], foreign_table: "contacts", foreign_columns: ["id"], on_delete: "no action", on_update: "no action" },
    }
}
